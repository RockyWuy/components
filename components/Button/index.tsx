import styled from "styled-components";
// import theme from '@themes/button/index.js';

interface ButtonProps {
  className?: string; // 样式类名
  disabled?: boolean; // 按钮失效状态 默认值 false
  loading?: boolean; // 设置按钮载入状态 默认值 false
  size?: string; // 设置按钮大小, 可选值为 small/large 或者不设
  type?: string; // 按钮类型, 可选值为 primary/ghost/warning 或者不设
  onClick?: () => void; // 点击按钮的回调
}

// 不同类型不同颜色
function typeToColor(
  type: string = "default",
  disabled: boolean = false,
  style: string
): string {
  let colors = {
    primary: {
      color: disabled ? "hsla(0,0%,100%,.6)" : "#fff",
      bgColor: "#108ee9",
      border: "1px solid #108ee9",
    },
    ghost: {
      color: disabled ? "rgba(0,0,0,.1)" : "#108ee9",
      bgColor: "transparent",
      border: disabled ? "1px solid rgba(0,0,0,.1)" : "1px solid #108ee9",
    },
    warning: {
      color: disabled ? "hsla(0,0%,100%,.6)" : "#fff",
      bgColor: "#e94f4f",
      border: "1px solid #e94f4f",
    },
    default: {
      color: disabled ? "rgba(0,0,0,.3)" : "#000",
      bgColor: "#fff",
      border: "1px solid #ddd",
    },
  };
  return colors[type][style];
}

// 激活时的颜色
function activeColor(type: string = "default", style: string): string {
  let colors = {
    primary: {
      color: "hsla(0,0%,100%,.3)",
      bgColor: "#0e80d2",
    },
    ghost: {
      color: "rgba(16,142,233,.6)",
      bgColor: "transparent",
      border: "1px solid rgba(16,142,233,.6)",
    },
    warning: {
      color: "hsla(0,0%,100%,.3)",
      bgColor: "#d24747",
    },
    default: {
      bgColor: "#ddd",
    },
  };
  return colors[type][style];
}

const Button = styled.a<ButtonProps>`
  display: block;
  outline: 0 none;
  box-sizing: border-box;
  padding: 0;
  text-align: center;
  font-size: ${(props) =>
    props.size === "small" ? "26px" : props.size === "large" ? "48px" : "36px"};
  font-weight: 400;
  height: 4.6rem;
  line-height: 4.6rem;

  color: ${(props) => typeToColor(props.type, props.disabled, "color")};
  background-color: ${(props) =>
    typeToColor(props.type, props.disabled, "bgColor")};
  border: ${(props) => typeToColor(props.type, props.disabled, "border")};
  opacity: ${(props) => (props.disabled ? 0.6 : 1)};
  border-radius: 0.5rem;

  overflow: hidden;
  text-overflow: ellipsis;
  word-break: break-word;
  white-space: nowrap;

  &:active {
    background-color: ${(props) =>
      props.disabled ? "" : activeColor(props.type, "bgColor")};
    color: ${(props) =>
      props.disabled ? "" : activeColor(props.type, "color")};
    border: ${(props) =>
      props.disabled ? "" : activeColor(props.type, "border")};
  }
`;

export default Button;
