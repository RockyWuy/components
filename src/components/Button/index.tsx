import styled from 'styled-components';

interface ButtonProps {
    disabled?: boolean;    // 按钮失效状态 默认值 false
    href?: string;         // 点击跳转的地址, 指定此属性 button 行为和 a 链接一致
    loading?: boolean;     // 设置按钮载入状态 默认值 false
    size?: string;         // 设置按钮大小, 可选值为 small/large 或者不设
    type?: string;         // 按钮类型, 可选值为 primary/dashed/danger/link 或者不设
    onClick?: () => void;  // 点击按钮的回调
}

const Button = styled.button<ButtonProps>`
    background: ${props => props.type ? 'blue' : 'white'};
    color: ${props => props.type ? 'white' : 'blue'};

    font-size: 1em;
    margin: 1em;
    padding: 0.25em 1em;
    border: 1px solid ${props => props.type ? 'white' : 'blue'};
    border-radius: 2px
`;

export default Button;