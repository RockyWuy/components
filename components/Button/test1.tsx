import * as React from "react";
import Button from "./index";

export default class TestButton extends React.Component<{}, {}> {
  public constructor(props: {}) {
    super(props);

    this.ClickTo = this.ClickTo.bind(this);
  }

  public ClickTo() {
    console.log("sss");
  }

  public render() {
    return (
      <React.Fragment>
        <Button className="btn" onClick={this.ClickTo}>
          default
        </Button>
        <div style={{ height: "10px" }}></div>
        <Button disabled>default</Button>
        <div style={{ height: "10px" }}></div>
        <Button type="primary">primary</Button>
        <div style={{ height: "10px" }}></div>
        <Button disabled type="primary">
          primary
        </Button>
        <div style={{ height: "10px" }}></div>
        <Button type="ghost">ghost</Button>
        <div style={{ height: "10px" }}></div>
        <Button disabled type="ghost">
          ghost
        </Button>
        <div style={{ height: "10px" }}></div>
        <Button type="warning">warning</Button>
        <div style={{ height: "10px" }}></div>
        <Button disabled type="warning">
          warning
        </Button>
      </React.Fragment>
    );
  }
}
