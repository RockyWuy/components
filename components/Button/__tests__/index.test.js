import * as React from 'react';
import { shallow, render } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';
import 'jest-styled-components';
import Button from '../index';

describe(`测试<Button> snapshots`, () => {
    const wrapper = shallow(<Button>sssa</Button>);
    const wrapper1 = render(<Button>sssa</Button>);
    it('render button', () => {
        expect(wrapper.props().children).toBe('sssa');
    });
    it('render button1', () => {
        expect(renderToJson(wrapper1)).toMatchSnapshot();
    });
});