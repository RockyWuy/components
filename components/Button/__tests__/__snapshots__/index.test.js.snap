// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试<Button> snapshots render button1 1`] = `
.c0 {
  display: block;
  outline: 0 none;
  box-sizing: border-box;
  padding: 0;
  text-align: center;
  font-size: 36px;
  font-weight: 400;
  height: 4.6rem;
  line-height: 4.6rem;
  color: #000;
  background-color: #fff;
  border: 1px solid #ddd;
  opacity: 1;
  border-radius: 0.5rem;
  overflow: hidden;
  text-overflow: ellipsis;
  word-break: break-word;
  white-space: nowrap;
}

.c0:active {
  background-color: #ddd;
}

<a
  class="c0"
>
  sssa
</a>
`;
